topics/integrate-in-existing-app.md

```diff
@@ -1,242 +1,415 @@
-    }
-    interface View {
-        fun displayModel(model: Model)
+       Log.i("Login Activity", "Hello from shared module: " + (Greeting().greeting()))
+   
     }
+    ```
 
-    private var view: View? = null
-    private var lastModel: Model = Model.Form
-    private fun displayModel(model: Model) {
-        lastModel = model
-        view?.displayModel(model)
-    }
+5. Search for `Hello` in the log, and you'll find the greeting from the shared module.
 
-    fun attach(view: View) {
-        this.view = view.apply { displayModel(lastModel) }
-    }
-    fun detach() { this.view = null }
+   ![Greeting from the shared module](shared-module-greeting.png)
 
-    sealed class Intent {
-        data class Login(val username: String, val password: String)
-            : Intent()
-    }
+### Make the business logic cross-platform
 
-    fun process(intent: Intent) {
-        when (intent) {
-            is Intent.Login -> {
-                MainScope().launch {
-                    try {
-                        displayModel(Model.Loading)
-                        getNetwork().login(
-                                intent.username, intent.password) // suspending
-                        displayModel(Model.GoToNextScreen)
-                    } catch (ex: LoginException) {
-                        displayModel(Model.Error(ex.code))
-                    }
-                }
-            }
-        }
+You can now extract the business logic code to the KMM shared module and make it platform-independent. This is necessary for reusing the code for both Android and iOS.
+
+1. Move the business logic code `com.jetbrains.simplelogin.androidapp.data` from the `app` directory to the `com.jetbrains.simplelogin.shared` package in the `shared/src/commonMain` directory.
+   You can drag and drop the package or refactor it by moving everything from one directory to another.
+
+   ![Drag and drop the package with the business logic code](moving-business-logic.png){width=350}
+
+2. When Android Studio asks what you'd like to do, select to move the package, and then approve the refactoring.
+
+   ![Refactor the business logic package](refactor-business-logic-package.png){width=500}
+
+3. Ignore all warnings about platform-dependent code and click **Continue**.
+
+   ![Warnings about platform-dependent code](warnings-android-specific-code.png){width=450}
+
+4. Remove Android-specific code by replacing it with cross-platform Kotlin code or connecting to Android-specific APIs using [`expect` and `actual` declarations](connect-to-platform-specific-apis.md). See the following sections for details.
+
+#### Replace Android-specific code with cross-platform code {initial-collapse-state="collapsed"}
+
+To make your code work well on both Android and iOS, replace all JVM dependencies with Kotlin dependencies wherever possible.
+
+1.  In the `login()` function of the `LoginDataSource` class, replace `IOException`, which is not available in Kotlin, with `RuntimeException`.
```

Comment:
minor: Some sentences (like this one) contain much inline code and can be rewritten to make them a little easier to read.
It's an idea of how to improve the user experience in the future.

Reply:
If you have ideas, please share. This variant came from copyeditors. I had a different one but it's not easy to read due to formatting, I agree. Maybe we could format inline code differently.